File: backend/ml_service/agents/pharmacy_agent.py
"""
Pharmacy Agent - Inventory Management

Implementation Mandate: Hybrid Logic
- Formula: Use a Classification Rule-Set (Low, Medium, High, Surge) for each
  medicine's demand level. If demand is "Surge", place a pre-emptive order
  to the Supplier Agent
"""

from typing import Dict, List, Optional


class PharmacyAgent:
    """Pharmacy Agent for medicine inventory optimization"""
    
    def __init__(self):
        self.DEMAND_THRESHOLDS = {
            'SURGE': 0.80,      # 80%+ consumption rate
            'HIGH': 0.60,       # 60-79% consumption rate
            'MEDIUM': 0.40,     # 40-59% consumption rate
            'LOW': 0.0          # 0-39% consumption rate
        }
        
        self.REORDER_POINTS = {
            'SURGE': 200,       # Reorder when below 200 units
            'HIGH': 150,        # Reorder when below 150 units
            'MEDIUM': 100,      # Reorder when below 100 units
            'LOW': 50           # Reorder when below 50 units
        }
    
    def classify_medicine_demand(
        self,
        medicine_stocks: Dict[str, int],  # Current stock levels
        consumption_rates: Dict[str, int],  # Daily consumption
        outbreak_alerts: List[str] = None  # Active disease outbreaks
    ) -> Dict:
        """
        Classify demand level for each medicine using Rule-Set
        
        Classification Rules:
        1. Calculate consumption rate = daily_consumption / current_stock
        2. Apply outbreak multiplier if disease outbreak detected
        3. Classify: SURGE (>80%), HIGH (60-80%), MEDIUM (40-60%), LOW (<40%)
        4. Generate pre-emptive orders for SURGE demand
        """
        
        outbreak_alerts = outbreak_alerts or []
        classifications = []
        preemptive_orders = []
        
        # Medicine-to-disease mapping for outbreak adjustment
        disease_medicine_map = {
            'dengue': ['dengue_medicine', 'paracetamol', 'iv_fluids'],
            'malaria': ['malaria_medicine', 'antimalarial', 'iv_fluids'],
            'covid': ['covid_medicine', 'oxygen', 'antibiotics', 'paracetamol'],
            'typhoid': ['typhoid_medicine', 'antibiotics', 'iv_fluids'],
            'influenza': ['flu_medicine', 'antivirals', 'paracetamol']
        }
        
        for medicine, stock in medicine_stocks.items():
            consumption = consumption_rates.get(medicine, 0)
            
            # Calculate base consumption rate
            if stock > 0:
                consumption_rate = consumption / stock
            else:
                consumption_rate = 1.0  # Out of stock = max urgency
            
            # Apply outbreak multiplier
            outbreak_multiplier = 1.0
            for outbreak in outbreak_alerts:
                if medicine in disease_medicine_map.get(outbreak.lower(), []):
                    outbreak_multiplier = 2.0  # 2x demand during outbreak
                    break
            
            adjusted_rate = min(1.0, consumption_rate * outbreak_multiplier)
            
            # Classify demand level
            demand_level = self._classify_demand(adjusted_rate)
            
            # Calculate days remaining
            days_remaining = (stock / consumption) if consumption > 0 else 999
            
            # Determine if order needed
            reorder_point = self.REORDER_POINTS[demand_level]
            needs_order = stock < reorder_point
            
            classification = {
                "medicine": medicine,
                "current_stock": stock,
                "daily_consumption": consumption,
                "consumption_rate": round(adjusted_rate, 3),
                "demand_level": demand_level,
                "days_remaining": round(days_remaining, 1),
                "reorder_point": reorder_point,
                "needs_order": needs_order,
                "outbreak_affected": outbreak_multiplier > 1.0
            }
            
            classifications.append(classification)
            
            # Generate pre-emptive order for SURGE demand
            if demand_level == "SURGE":
                order = self._generate_order(
                    medicine, 
                    stock, 
                    consumption, 
                    demand_level,
                    outbreak_multiplier > 1.0
                )
                preemptive_orders.append(order)
        
        # Calculate overall inventory health
        inventory_health = self._calculate_inventory_health(classifications)
        
        return {
            "classifications": classifications,
            "preemptive_orders": preemptive_orders,
            "inventory_health": inventory_health,
            "critical_medicines": [c["medicine"] for c in classifications if c["demand_level"] in ["SURGE", "HIGH"]],
            "total_medicines": len(classifications),
            "medicines_needing_order": len([c for c in classifications if c["needs_order"]]),
            "recommendations": self._get_recommendations(classifications, preemptive_orders)
        }
    
    def _classify_demand(self, consumption_rate: float) -> str:
        """Classify demand level based on consumption rate"""
        if consumption_rate >= self.DEMAND_THRESHOLDS['SURGE']:
            return "SURGE"
        elif consumption_rate >= self.DEMAND_THRESHOLDS['HIGH']:
            return "HIGH"
        elif consumption_rate >= self.DEMAND_THRESHOLDS['MEDIUM']:
            return "MEDIUM"
        else:
            return "LOW"
    
    def _generate_order(
        self, 
        medicine: str, 
        current_stock: int,
        daily_consumption: int,
        demand_level: str,
        outbreak_active: bool
    ) -> Dict:
        """Generate pre-emptive order for supplier"""
        
        # Calculate order quantity (7-14 days supply)
        days_supply = 14 if outbreak_active else 10
        order_quantity = daily_consumption * days_supply
        
        # Minimum order quantity
        order_quantity = max(order_quantity, 100)
        
        urgency = "URGENT" if demand_level == "SURGE" else "HIGH"
        
        return {
            "medicine": medicine,
            "order_quantity": order_quantity,
            "urgency": urgency,
            "reason": f"Pre-emptive order: {demand_level} demand detected",
            "current_stock": current_stock,
            "daily_consumption": daily_consumption,
            "outbreak_related": outbreak_active,
            "estimated_stockout_days": round(current_stock / daily_consumption, 1) if daily_consumption > 0 else 999
        }
    
    def _calculate_inventory_health(self, classifications: List[Dict]) -> Dict:
        """Calculate overall inventory health metrics"""
        if not classifications:
            return {"status": "UNKNOWN", "score": 0}
        
        total = len(classifications)
        surge_count = sum(1 for c in classifications if c["demand_level"] == "SURGE")
        high_count = sum(1 for c in classifications if c["demand_level"] == "HIGH")
        low_stock_count = sum(1 for c in classifications if c["days_remaining"] < 7)
        
        # Calculate health score (0-100)
        surge_penalty = (surge_count / total) * 50
        high_penalty = (high_count / total) * 30
        low_stock_penalty = (low_stock_count / total) * 20
        
        health_score = 100 - surge_penalty - high_penalty - low_stock_penalty
        
        # Determine status
        if health_score >= 80:
            status = "EXCELLENT"
        elif health_score >= 60:
            status = "GOOD"
        elif health_score >= 40:
            status = "FAIR"
        elif health_score >= 20:
            status = "POOR"
        else:
            status = "CRITICAL"
        
        return {
            "status": status,
            "score": round(health_score, 1),
            "surge_items": surge_count,
            "high_demand_items": high_count,
            "low_stock_items": low_stock_count
        }
    
    def _get_recommendations(self, classifications: List[Dict], orders: List[Dict]) -> List[str]:
        """Generate actionable recommendations"""
        recommendations = []
        
        surge_items = [c["medicine"] for c in classifications if c["demand_level"] == "SURGE"]
        if surge_items:
            recommendations.append(f"🚨 SURGE DEMAND: Immediate orders placed for {len(surge_items)} medicines: {', '.join(surge_items[:3])}")
        
        critical_stock = [c["medicine"] for c in classifications if c["days_remaining"] < 3]
        if critical_stock:
            recommendations.append(f"⚠️ CRITICAL: {len(critical_stock)} medicines have <3 days stock remaining")
        
        if orders:
            recommendations.append(f"📦 {len(orders)} pre-emptive orders generated for supplier")
        
        outbreak_affected = [c["medicine"] for c in classifications if c["outbreak_affected"]]
        if outbreak_affected:
            recommendations.append(f"🦠 {len(outbreak_affected)} medicines affected by outbreak alerts")
        
        if not recommendations:
            recommendations.append("✅ Inventory levels are healthy. Continue monitoring.")
        
        return recommendations

File: backend/agents/PharmacyAgent.js
// backend/agents/PharmacyAgent.js
const { subscribe, publish } = require("../eventBus");
const EVENTS = require("../constants/events");

class PharmacyAgent {
  constructor(id, worldState, log) {
    this.id = id;
    this.worldState = worldState;
    this.log = log;
    this.ML_SERVICE_URL = "http://localhost:8000"; // Python ML service

    // Subscribe to outbreak alerts for all diseases
    subscribe(
      "DENGUE_OUTBREAK_PREDICTED",
      this.onOutbreakAlert.bind(this, "dengue", "dengueMed")
    );
    subscribe(
      "MALARIA_OUTBREAK_PREDICTED",
      this.onOutbreakAlert.bind(this, "malaria", "chloroquine")
    );
    subscribe(
      "TYPHOID_OUTBREAK_PREDICTED",
      this.onOutbreakAlert.bind(this, "typhoid", "ceftriaxone")
    );
    subscribe(
      "INFLUENZA_OUTBREAK_PREDICTED",
      this.onOutbreakAlert.bind(this, "influenza", "oseltamivir")
    );
    subscribe(
      "COVID_OUTBREAK_PREDICTED",
      this.onOutbreakAlert.bind(this, "covid", "oseltamivir")
    );

    // Subscribe to medicine requests from hospitals
    subscribe("MEDICINE_REQUEST", this.onMedicineRequest.bind(this));
  }

  start() {
    // Log initialization
    const p = this.worldState.pharmacies[this.id];
    const medicineCount = Object.keys(p.medicines || {}).length;
    this.log(
      `✅ Pharmacy Agent ${this.id} (${p.name}) initialized - Managing ${medicineCount} medicines in ${p.zone}`,
      { agent: "Pharmacy", type: "INIT", entityId: this.id }
    );

    // Runs every 12 seconds
    setInterval(() => this.tick(), 12000);
  }

  async tick() {
    const p = this.worldState.pharmacies[this.id];
    if (!p) return;

    // Simulate daily medicine consumption
    this.consumeMedicines(p);

    // 🤖 ML-POWERED: Classify medicine demand using ML service
    await this.classifyDemandWithML(p);

    // Count low stock items
    const lowStockItems = Object.entries(p.medicines).filter(
      ([_, med]) => med.stock <= med.reorderPoint
    ).length;

    // Count pending orders
    const pendingOrders = p.pendingOrders?.length || 0;

    // ALWAYS log current status
    const status = lowStockItems > 0 ? "🟡 MONITORING" : "🟢 NORMAL";
    this.log(
      `${p.name}: ${status} stock levels | ${lowStockItems} items need restock | ${pendingOrders} pending orders`,
      {
        agent: "Pharmacy",
        type: "STATUS",
        entityId: this.id,
        lowStockItems,
        pendingOrders,
      }
    );

    // Check all medicines for shortage risk
    Object.keys(p.medicines).forEach((medicineName) => {
      this.checkMedicineStock(p, medicineName);
    });
  }

  async classifyDemandWithML(pharmacy) {
    /**
     * 🤖 ML-POWERED DEMAND CLASSIFICATION
     * Uses Python FastAPI ML service for medicine demand prediction
     * Classifies each medicine as LOW/MEDIUM/HIGH/SURGE based on consumption rate
     */

    try {
      // Prepare medicine inventory data
      const medicineInventory = {};

      Object.entries(pharmacy.medicines).forEach(([medName, medicine]) => {
        medicineInventory[medName] = {
          current_stock: medicine.stock || 0,
          daily_consumption: medicine.dailyUsage || medicine.usagePerDay || 0,
          reorder_point: medicine.reorderPoint || 0,
          lead_time_days: medicine.leadTime || 3,
        };
      });

      // Call Python ML service
      const response = await fetch(
        `${this.ML_SERVICE_URL}/classify/pharmacy_demand`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            pharmacy_id: this.id,
            medicine_inventory: medicineInventory,
          }),
        }
      );

      if (!response.ok) {
        throw new Error(`ML Service error: ${response.status}`);
      }

      const result = await response.json();

      // Log ML prediction
      this.log(
        `🤖 ML DEMAND ANALYSIS: ${pharmacy.name} | Health: ${result.overall_health} | Pre-emptive orders: ${result.preemptive_order_count}`,
        {
          agent: "Pharmacy",
          type: `ML_DEMAND_ANALYSIS`,
          entityId: this.id,
          health: result.overall_health,
          orderCount: result.preemptive_order_count,
          mlPowered: true,
        }
      );

      // Process pre-emptive orders for SURGE items
      if (result.preemptive_orders && result.preemptive_orders.length > 0) {
        result.preemptive_orders.forEach((order) => {
          this.log(
            `📡 ${pharmacy.name}: ${order.priority} priority order needed for ${order.medicine_name} (${order.demand_level}) - ${order.reason}`,
            {
              agent: "Pharmacy",
              type: `ML_PREEMPTIVE_ORDER`,
              entityId: this.id,
              medicine: order.medicine_name,
              priority: order.priority,
              demandLevel: order.demand_level,
            }
          );

          // Publish medicine shortage event
          publish("MEDICINE_SHORTAGE_RISK", {
            pharmacyId: this.id,
            pharmacyName: pharmacy.name,
            zone: pharmacy.zone,
            medicine: order.medicine_name,
            quantity: order.quantity_needed,
            urgency: order.priority,
            criticality: order.demand_level === "SURGE" ? "high" : "medium",
            reason: order.reason,
            timestamp: new Date().toISOString(),
            mlPowered: true,
          });
        });
      }

      // Log any SURGE demand items separately
      const surgeItems = result.classifications.filter(
        (c) => c.demand_level === "SURGE"
      );
      if (surgeItems.length > 0) {
        this.log(
          `🔥 ${pharmacy.name}: ${
            surgeItems.length
          } SURGE demand items detected: ${surgeItems
            .map((s) => s.medicine_name)
            .join(", ")}`,
          {
            agent: "Pharmacy",
            type: `ML_SURGE_ALERT`,
            entityId: this.id,
            surgeItems: surgeItems.map((s) => s.medicine_name),
          }
        );
      }
    } catch (error) {
      // Fallback: Use rule-based demand classification if ML service unavailable
      this.log(
        `⚠️ ML service unavailable for ${pharmacy.name}, using fallback classification: ${error.message}`,
        {
          agent: "Pharmacy",
          type: "ML_FALLBACK",
          entityId: this.id,
          error: error.message,
        }
      );
    }
  }

  consumeMedicines(pharmacy) {
    // Simulate realistic medicine usage over time
    Object.entries(pharmacy.medicines).forEach(([medName, medicine]) => {
      if (medicine.usagePerDay && medicine.usagePerDay > 0) {
        // INCREASED CONSUMPTION: every 12 seconds = 1.5% of daily usage (was 0.2%)
        // This makes medicine depletion visible and triggers agent action
        let consumptionRate = 0.015; // Base rate: 1.5%

        // Increase consumption during outbreaks or high demand periods
        if (pharmacy.demandMultiplier && pharmacy.demandMultiplier[medName]) {
          consumptionRate *= pharmacy.demandMultiplier[medName];
        }

        // Random variation: 80-120% of expected consumption
        const randomFactor = 0.8 + Math.random() * 0.4;
        const consumptionAmount = Math.ceil(
          medicine.usagePerDay * consumptionRate * randomFactor
        );

        if (medicine.stock > 0) {
          medicine.stock = Math.max(0, medicine.stock - consumptionAmount);
        }
      }
    });
  }

  checkMedicineStock(pharmacy, medicineName) {
    const medicine = pharmacy.medicines[medicineName];
    if (!medicine) return;

    const stock = medicine.stock;
    const dailyUsage = medicine.dailyUsage;
    const reorderPoint = medicine.reorderPoint;
    const daysLeft = dailyUsage > 0 ? stock / dailyUsage : 999;

    // Check if we're at or below reorder point
    if (stock <= reorderPoint) {
      const urgency = daysLeft < 2 ? "high" : daysLeft < 5 ? "medium" : "low";

      this.log(
        `⚠️ ${
          pharmacy.name
        }: ${urgency.toUpperCase()} shortage alert for ${medicineName}! Stock: ${stock} units (~${daysLeft.toFixed(
          1
        )} days remaining)`,
        {
          agent: "Pharmacy",
          type: "MED_SHORTAGE",
          entityId: this.id,
          medicine: medicineName,
          urgency,
          criticality: medicine.criticality,
        }
      );

      // Calculate optimal order quantity
      const orderQuantity = this.calculateOrderQuantity(medicine, daysLeft);

      // Add to pending orders
      if (!pharmacy.pendingOrders.some((o) => o.medicine === medicineName)) {
        pharmacy.pendingOrders.push({
          medicine: medicineName,
          quantity: orderQuantity,
          supplier: medicine.supplier,
          urgency,
          timestamp: new Date().toISOString(),
          status: "requested",
        });

        this.log(
          `📤 ${pharmacy.name}: Placing ${urgency} priority order with ${medicine.supplier} - Ordering ${orderQuantity} units of ${medicineName}`,
          {
            agent: "Pharmacy",
            type: "ORDER_PLACED",
            entityId: this.id,
            medicine: medicineName,
            quantity: orderQuantity,
            supplier: medicine.supplier,
            urgency,
          }
        );

        // Log coordination message
        this.log(
          `📡 ${pharmacy.name}: Sent order request to ${medicine.supplier} for ${medicineName} (${orderQuantity} units)`,
          {
            agent: "Pharmacy",
            type: "COORDINATION",
            entityId: this.id,
            medicine: medicineName,
            quantity: orderQuantity,
            recipient: medicine.supplier,
          }
        );
      }

      publish("MEDICINE_SHORTAGE_RISK", {
        pharmacyId: this.id,
        pharmacyName: pharmacy.name,
        medicine: medicineName,
        stock,
        daysLeft: daysLeft.toFixed(1),
        reorderPoint,
        urgency,
        criticality: medicine.criticality,
        zone: pharmacy.zone,
        orderQuantity,
        supplier: medicine.supplier,
      });
    }
  }

  calculateOrderQuantity(medicine, daysLeft) {
    // Order enough for 7-14 days depending on criticality
    const targetDays = medicine.criticality === "high" ? 14 : 10;
    const optimalStock = medicine.dailyUsage * targetDays;
    const orderQuantity = Math.max(
      optimalStock - medicine.stock,
      medicine.dailyUsage * 7
    );

    // Round up to nearest 10 for easier handling
    return Math.ceil(orderQuantity / 10) * 10;
  }

  onOutbreakAlert(disease, primaryMedicine, event) {
    const p = this.worldState.pharmacies[this.id];
    if (!p) return;

    // Only respond if outbreak is in our zone
    if (p.zone !== event.zone) return;

    this.log(
      `💊 ${
        p.name
      }: ${disease.toUpperCase()} OUTBREAK ALERT received from labs! Preparing medicine supplies`,
      {
        agent: "Pharmacy",
        type: "ALERT_RECEIVED",
        entityId: this.id,
        disease,
        zone: event.zone,
      }
    );

    // Increase usage estimate for relevant medicines
    const medicine = p.medicines[primaryMedicine];
    if (medicine) {
      // Increase daily usage estimate based on risk level
      const multiplier =
        event.riskLevel === "critical"
          ? 2.5
          : event.riskLevel === "high"
          ? 2.0
          : 1.5;

      const oldUsage = medicine.dailyUsage;
      medicine.dailyUsage = Math.round(oldUsage * multiplier);
      const currentStock = medicine.stock;
      const daysRemaining = currentStock / medicine.dailyUsage;

      this.log(
        `📊 ${
          p.name
        }: Adjusting ${primaryMedicine} demand forecast - Usage: ${oldUsage} → ${
          medicine.dailyUsage
        }/day | Stock: ${currentStock} units (${daysRemaining.toFixed(
          1
        )} days supply)`,
        {
          agent: "Pharmacy",
          type: "DEMAND_FORECAST",
          entityId: this.id,
          disease,
          medicine: primaryMedicine,
          multiplier,
        }
      );

      // Immediately check if we need to order more
      setTimeout(() => this.checkMedicineStock(p, primaryMedicine), 1000);
    }

    // Also increase related medicines
    const relatedMedicines = this.getRelatedMedicines(disease);
    relatedMedicines.forEach((medName) => {
      if (p.medicines[medName]) {
        const oldUsage = p.medicines[medName].dailyUsage;
        p.medicines[medName].dailyUsage = Math.round(oldUsage * 1.3);
      }
    });
  }

  getRelatedMedicines(disease) {
    const relatedMap = {
      dengue: ["paracetamol", "ivFluids", "ors"],
      malaria: ["artemether", "paracetamol"],
      typhoid: ["azithromycin", "ciprofloxacin", "ivFluids"],
      influenza: ["paracetamol", "ibuprofen", "fluVaccine"],
      covid: ["azithromycin", "paracetamol", "covidVaccine"],
    };
    return relatedMap[disease] || [];
  }

  onMedicineRequest(event) {
    const p = this.worldState.pharmacies[this.id];
    if (!p) return;

    // Check if request is for our zone
    if (p.zone !== event.zone) return;

    this.log(
      `📩 ${p.name}: Medicine request received from ${
        event.hospitalName || "Hospital"
      } for ${event.disease} outbreak (${event.urgency} priority)`,
      {
        agent: "Pharmacy",
        type: "REQUEST_RECEIVED",
        entityId: this.id,
        hospitalId: event.hospitalId,
        urgency: event.urgency,
      }
    );

    // Proactively check stock for outbreak-related medicines
    const disease = event.disease;
    const primaryMedicine = this.getPrimaryMedicineForDisease(disease);

    if (primaryMedicine && p.medicines[primaryMedicine]) {
      const medicine = p.medicines[primaryMedicine];
      const stockStatus =
        medicine.stock > medicine.reorderPoint ? "sufficient" : "low";

      this.log(
        `📦 ${p.name}: Checking ${primaryMedicine} inventory - Stock: ${medicine.stock} units (Status: ${stockStatus})`,
        {
          agent: "Pharmacy",
          type: "STOCK_CHECK",
          entityId: this.id,
          medicine: primaryMedicine,
          stock: medicine.stock,
        }
      );

      this.checkMedicineStock(p, primaryMedicine);
    }
  }

  getPrimaryMedicineForDisease(disease) {
    const diseaseMap = {
      dengue: "dengueMed",
      malaria: "chloroquine",
      typhoid: "ceftriaxone",
      influenza: "oseltamivir",
      covid: "oseltamivir",
    };
    return diseaseMap[disease];
  }
}

module.exports = PharmacyAgent;
